File: sdk/web/scripts/test-sqlite3mc-app-access.js
import assert from "node:assert/strict";

export async function run(ctx) {
  const password = "synthetic application password";
  const wait = expression => ctx.js(`for(let i=0;i<300;i++){if(${expression})return true;await new Promise(r=>setTimeout(r,30));}throw Error('UI condition timed out');`);
  async function page(setup = false) {
    await ctx.load();
    await ctx.js("if(arguments[0])history.replaceState(null,'','?storage=encrypted');window.ready=false;await new Promise((resolve,reject)=>{const s=document.createElement('script');s.type='module';s.src='/test-access-entry.js';s.onload=resolve;s.onerror=()=>reject(Error('access entry failed'));document.head.append(s);});window.store=new keyModule.IndexedDbKeyStore();accessUi.startStorageAccess().then(()=>window.ready=true);return true;", [setup]);
    await wait("document.querySelector('#storage-create') && (document.querySelector('dialog').open || window.ready)");
  }
  async function click(name, passwordValue = password, extra = "") {
    await ctx.js(`document.querySelector('#storage-password').value=arguments[0];${extra}document.querySelector('#storage-${name}').click();return true;`, [passwordValue]);
    await wait(`!document.querySelector('#storage-${name}').disabled`);
  }
  await page(true);
  if (ctx.args.plaintext_only) {
    assert(await ctx.js("return !ready && document.querySelector('#storage-create').disabled && document.querySelector('#storage-feedback').textContent.includes('cannot open encrypted');"));
    assert.deepEqual(await ctx.snapshot(), ctx.legacy);
    ctx.checks.push("stock build blocks encrypted setup before creating key or database");
    await page(); assert(await ctx.js("return ready;"));
    ctx.checks.push("stock build retains default plaintext startup");
    return;
  }
  assert(!(await ctx.js("return ready;")));
  await click("create", password, "document.querySelector('#storage-confirm').value=arguments[0];");
  assert(await ctx.js("return !ready && (await store.read('spp.encrypted.db'))!==null && localStorage.getItem('spp.storage-access.v1')==='backup-required' && document.querySelector('#storage-resume').hidden && document.querySelector('#storage-manage').hidden;"));
  ctx.checks.push("setup UI persists wrapped key and creates separate encrypted OPFS before app initialization");
  await click("unlock");
  assert(await ctx.js("return !ready && document.querySelector('#storage-feedback').textContent.includes('backup');"));
  await ctx.js("window.backupText=null;URL.createObjectURL=blob=>{blob.text().then(text=>window.backupText=text);return 'blob:synthetic';};HTMLAnchorElement.prototype.click=function(){};return true;");
  await click("backup"); await wait("window.backupText!==null");
  await ctx.js("sessionStorage.setItem('test-backup',backupText);return true;");
  await click("unlock", "wrong");
  assert(await ctx.js("return !ready && document.querySelector('#storage-feedback').textContent.includes('Incorrect password');"));
  await click("unlock"); await wait("window.ready");
  await ctx.js("window.appStorage=await facade.ensureStorage();await appStorage.setSetting('app-access-protected',arguments[0]);return true;", [ctx.marker]);
  ctx.checks.push("backup required during setup; wrong password rejected; correct password opens app facade");
  await ctx.js("setTimeout(()=>{[...document.querySelectorAll('button')].find(b=>b.textContent==='Lock database').click();},50);return true;");
  await new Promise(resolve => setTimeout(resolve, 500));
  await page(); assert(!(await ctx.js("return ready;")));
  await ctx.js("const input=document.querySelector('#storage-password');input.value=arguments[0];input.focus();return true;", [password]);
  await ctx.request("POST", `/session/${ctx.sid}/actions`, { actions: [{ type: "key", id: "keyboard", actions: [{ type: "keyDown", value: "\ue007" }, { type: "keyUp", value: "\ue007" }] }] });
  await wait("window.ready");
  assert.equal(await ctx.js("return await (await facade.ensureStorage()).getSetting('app-access-protected');"), ctx.marker);
  await ctx.js("await facade.closeStorageForLock();return true;");
  ctx.checks.push("click unlock and Enter unlock both work; lock/reload preserve encrypted application settings");
  if (!ctx.firefox) {
    const cdp = (command, params) => ctx.request("POST", `/session/${ctx.sid}/goog/cdp/execute`, { cmd: command, params });
    await cdp("WebAuthn.enable", {});
    const authenticator = (await cdp("WebAuthn.addVirtualAuthenticator", { options: { protocol: "ctap2", ctap2Version: "ctap2_1", transport: "internal", hasResidentKey: true, hasUserVerification: true, isUserVerified: true, automaticPresenceSimulation: true, hasPrf: true } })).authenticatorId;
    await ctx.js("[...document.querySelectorAll('button')].find(b=>b.textContent==='Database security').click();return true;");
    await click("enroll");
    assert(await ctx.js("return (await store.read('spp.encrypted.db')).passkey!==null;"));
    await page(); await click("passkey", ""); await wait("window.ready");
    assert.equal(await ctx.js("return await (await facade.ensureStorage()).getSetting('app-access-protected');"), ctx.marker);
    await ctx.js("await facade.closeStorageForLock();return true;");
    await page();
    const replacement = "synthetic replacement password";
    await click("recover", "", `document.querySelector('#storage-new').value='${replacement}';document.querySelector('#storage-new-confirm').value='${replacement}';`);
    assert(await ctx.js("return !ready && document.querySelector('#storage-feedback').textContent.includes('Password reset');"));
    await click("unlock", replacement); await wait("window.ready");
    await ctx.js("await facade.closeStorageForLock();return true;");
    await cdp("WebAuthn.removeVirtualAuthenticator", { authenticatorId: authenticator });
    ctx.checks.push("actual Chromium WebAuthn with virtual PRF authenticator enrolls in UI, unlocks after reload and resets forgotten password while locked");
  }
  await ctx.js("await new Promise((resolve,reject)=>{const r=indexedDB.deleteDatabase('spp-database-key-vault');r.onsuccess=resolve;r.onerror=()=>reject(r.error);});return true;");
  await page();
  assert(await ctx.js("return !ready && !document.querySelector('#storage-restore-section').hidden && document.querySelector('#storage-create').hidden;"));
  await ctx.js("const transfer=new DataTransfer();transfer.items.add(new File([sessionStorage.getItem('test-backup')],'backup.json',{type:'application/json'}));document.querySelector('#storage-file').files=transfer.files;return true;");
  const before = await ctx.snapshot();
  await click("restore", "wrong");
  assert(await ctx.js("return (await store.read('spp.encrypted.db'))===null && !ready;"));
  assert.deepEqual(await ctx.snapshot(), before);
  await click("restore");
  assert(await ctx.js("return (await store.read('spp.encrypted.db'))!==null && !ready;"));
  await click("unlock"); await wait("window.ready");
  assert.equal(await ctx.js("return await (await facade.ensureStorage()).getSetting('app-access-protected');"), ctx.marker);
  await ctx.js("await facade.closeStorageForLock();return true;");
  ctx.checks.push("backup recovery rejects wrong password without mutation and restores access to existing encrypted data");
  const final = await ctx.snapshot();
  assert.deepEqual(final.filter(file => !file.path.startsWith(".opfs-sahpool-encrypted/")), ctx.legacy);
  assert(!final.some(file => file.protected));
  ctx.checks.push("new setup leaves original plaintext database byte-identical and protected data absent from artifacts");
}

File: sdk/web/scripts/test-sqlite3mc-app-access.py
"""Application access UI with real key wrapping, facade, IndexedDB and encrypted OPFS."""


def run(ctx):
    js, checks = ctx['js'], ctx['checks']
    password = 'synthetic application password'

    def page(setup=False):
        ctx['load']()
        js("if(arguments[0])history.replaceState(null,'','?storage=encrypted');window.ready=false;await new Promise((resolve,reject)=>{const s=document.createElement('script');s.type='module';s.src='/test-access-entry.js';s.onload=resolve;s.onerror=()=>reject(Error('access entry failed'));document.head.append(s);});window.store=new keyModule.IndexedDbKeyStore();accessUi.startStorageAccess().then(()=>window.ready=true);return true;", [setup])
        wait("document.querySelector('#storage-create') && (document.querySelector('dialog').open || window.ready)")

    def wait(expression):
        js("for(let i=0;i<300;i++){if(" + expression + ")return true;await new Promise(r=>setTimeout(r,30));}throw Error('UI condition timed out');")

    def click(name, password_value=password, extra=''):
        js("document.querySelector('#storage-password').value=arguments[0];" + extra + "document.querySelector('#storage-" + name + "').click();return true;", [password_value])
        wait("!document.querySelector('#storage-" + name + "').disabled")

    page(True)
    if ctx['args'].plaintext_only:
        assert js("return !ready && document.querySelector('#storage-create').disabled && document.querySelector('#storage-feedback').textContent.includes('cannot open encrypted');")
        assert ctx['snapshot']() == ctx['legacy']
        checks.append('stock build blocks encrypted setup before creating key or database')
        page()
        assert js('return ready;')
        checks.append('stock build retains default plaintext startup')
        return

    assert js('return !ready;')
    click('create', extra="document.querySelector('#storage-confirm').value=arguments[0];")
    assert js("return !ready && (await store.read('spp.encrypted.db'))!==null && localStorage.getItem('spp.storage-access.v1')==='backup-required' && document.querySelector('#storage-resume').hidden && document.querySelector('#storage-manage').hidden;")
    checks.append('setup UI persists wrapped key and creates separate encrypted OPFS before app initialization')
    click('unlock')
    assert js("return !ready && document.querySelector('#storage-feedback').textContent.includes('backup');")
    # Capture download content in this synthetic profile, without returning key metadata to the driver.
    js("window.backupText=null;URL.createObjectURL=blob=>{blob.text().then(text=>window.backupText=text);return 'blob:synthetic';};HTMLAnchorElement.prototype.click=function(){};return true;")
    click('backup')
    wait('window.backupText!==null')
    js("sessionStorage.setItem('test-backup',backupText);return true;")
    click('unlock', 'wrong')
    assert js("return !ready && document.querySelector('#storage-feedback').textContent.includes('Incorrect password');")
    click('unlock')
    wait('window.ready')
    js("window.appStorage=await facade.ensureStorage();await appStorage.setSetting('app-access-protected',arguments[0]);return true;", [ctx['marker']])
    checks.append('backup required during setup; wrong password rejected; correct password opens app facade')
    # Actual lock path closes the shared worker then reloads; wait through navigation using WebDriver.
    js("setTimeout(()=>{[...document.querySelectorAll('button')].find(b=>b.textContent==='Lock database').click();},50);return true;")
    import time
    time.sleep(.5)
    page()
    assert js('return !ready;')
    js("const input=document.querySelector('#storage-password');input.value=arguments[0];input.focus();return true;", [password])
    ctx['request']('POST', f"/session/{ctx['sid']}/actions", {'actions': [{
        'type': 'key', 'id': 'keyboard', 'actions': [
            {'type': 'keyDown', 'value': '\ue007'}, {'type': 'keyUp', 'value': '\ue007'},
        ],
    }]})
    wait('window.ready')
    assert js("return await (await facade.ensureStorage()).getSetting('app-access-protected');") == ctx['marker']
    js('await facade.closeStorageForLock();return true;')
    checks.append('click unlock and Enter unlock both work; lock/reload preserve encrypted application settings')

    if not ctx['firefox']:
        def cdp(command, params):
            return ctx['request']('POST', f"/session/{ctx['sid']}/goog/cdp/execute", {'cmd': command, 'params': params})
        cdp('WebAuthn.enable', {})
        authenticator = cdp('WebAuthn.addVirtualAuthenticator', {'options': {
            'protocol': 'ctap2', 'ctap2Version': 'ctap2_1', 'transport': 'internal',
            'hasResidentKey': True, 'hasUserVerification': True, 'isUserVerified': True,
            'automaticPresenceSimulation': True, 'hasPrf': True,
        }})['authenticatorId']
        js("[...document.querySelectorAll('button')].find(b=>b.textContent==='Database security').click();return true;")
        click('enroll')
        assert js("return (await store.read('spp.encrypted.db')).passkey!==null;")
        page()
        click('passkey', '')
        wait('window.ready')
        assert js("return await (await facade.ensureStorage()).getSetting('app-access-protected');") == ctx['marker']
        js('await facade.closeStorageForLock();return true;')
        page()
        replacement = 'synthetic replacement password'
        click('recover', '', "document.querySelector('#storage-new').value='" + replacement + "';document.querySelector('#storage-new-confirm').value='" + replacement + "';")
        assert js("return !ready && document.querySelector('#storage-feedback').textContent.includes('Password reset');")
        click('unlock', replacement)
        wait('window.ready')
        js('await facade.closeStorageForLock();return true;')
        cdp('WebAuthn.removeVirtualAuthenticator', {'authenticatorId': authenticator})
        checks.append('actual Chromium WebAuthn with virtual PRF authenticator enrolls in UI, unlocks after reload and resets forgotten password while locked')

    # Simulate loss of wrapping metadata only, never delete OPFS or the encrypted mode marker.
    js("await new Promise((resolve,reject)=>{const r=indexedDB.deleteDatabase('spp-database-key-vault');r.onsuccess=resolve;r.onerror=()=>reject(r.error);});return true;")
    page()
    assert js("return !ready && !document.querySelector('#storage-restore-section').hidden && document.querySelector('#storage-create').hidden;")
    js("const transfer=new DataTransfer();transfer.items.add(new File([sessionStorage.getItem('test-backup')],'backup.json',{type:'application/json'}));document.querySelector('#storage-file').files=transfer.files;return true;")
    before = ctx['snapshot']()
    click('restore', 'wrong')
    assert js("return (await store.read('spp.encrypted.db'))===null && !ready;")
    assert ctx['snapshot']() == before
    click('restore')
    assert js("return (await store.read('spp.encrypted.db'))!==null && !ready;")
    click('unlock')
    wait('window.ready')
    assert js("return await (await facade.ensureStorage()).getSetting('app-access-protected');") == ctx['marker']
    js('await facade.closeStorageForLock();return true;')
    checks.append('backup recovery rejects wrong password without mutation and restores access to existing encrypted data')
    final = ctx['snapshot']()
    assert [f for f in final if not f['path'].startswith('.opfs-sahpool-encrypted/')] == ctx['legacy']
    assert not any(f.get('protected') for f in final)
    checks.append('new setup leaves original plaintext database byte-identical and protected data absent from artifacts')
